Plumb initial translation of `thread.spawn_indirect`

abrown opened the pull request:
This change in no way implements actual spawning of threads; it simply translates the CM builtin `thread.spawn_indirect` to invoke a real Wasmtime libcall that does some preliminary checks and panics. It is useful, though, to (1) exercise all the new table-related functions and (2) show the limits of Wasmtime's current state: we will need a way of interacting with shared functions and possibly some form of shared store.

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

abrown committed:
Plumb initial translation of `thread.spawn_indirect`

This change in no way implements actual spawning of threads; it simply
translates the CM builtin `thread.spawn_indirect` to invoke a real
Wasmtime libcall that does some preliminary checks and panics. It is
useful, though, to (1) exercise all the new table-related functions and
(2) show the limits of Wasmtime's current state: we will need a way of
interacting with shared functions and possibly some form of shared
store.

abrown commented on crates/cranelift/src/compiler/component.rs:
```diff
@@ -280,6 +280,11 @@ impl<'a> TrampolineCompiler<'a> {
+            Trampoline::ThreadSpawnIndirect { ty: _, table } => {
+                // TODO: eventually pass through the `ty` argument to check the
+                // table's funcref signature.
+                self.translate_thread_spawn_indirect(*table)
```
@alexcrichton, I feel like there is some way to actually materialize this `ty` index into a real type so we can check that the `funcref` has the expected type of the thread start function but I don't see it.

alexcrichton replied:
`TypeFuncIndex` is fine yeah, and the way to pass it to runtime would basically be to embed the constant in the generated code itself. Runtime would then be responsible for translating `TypeFuncIndex` into a `VMSharedTypeIndex` which can be done through various translation tables similar to how core wasm does it

abrown replied:
Ok, but actually the index being used is coming from a core function type. E.g., in the tests, for `(core type $start (shared (func (param $context i32))))`, we use `$start` as the index passed through to this canonical built-in. I see how to retrieve a `TypeFunc` using `ComponentTypes` but I don't think that's what we actually want here &mdash; that would be a component type. Is `TypeFuncIndex` the right index to be using here? How do we get to its core type (presumably backed by `VMSharedTypeIndex`)?

alexcrichton replied:
Agreed yeah we don't want to deal with `TypeFunc` here as the type-check should be a single comparison ideally, via `VMShareTypeIndex`. Converting from `TypeFuncIndex` will require a table of some form stored in `VMComponentContext` which is filled in on initialization. That happens for core modules, for example, where basically `VMContext` has a table from `TypeFuncIndex` to `VMSharedTypeIndex`

abrown commented on crates/environ/src/component.rs:
```diff
@@ -154,6 +154,9 @@ macro_rules! foreach_builtin_component_function {
             error_context_transfer(vmctx: vmctx, src_idx: u32, src_table: u32, dst_table: u32) -> u64;
 
+            #[cfg(feature = "threads")]
+            thread_spawn_indirect(vmctx: vmctx, table: u32, element: u32, context: u32 ) -> u64;
```
For 64-bit WebAssembly, we probably want `element` (and probably even `context`) to be a `u64`; how are we handling the 32-bit/64-bit distinction elsewhere?

alexcrichton replied:
Currently it's basically not-handled, Wasmtime [doesn't support 64-bit components at all](https://github.com/bytecodealliance/wasmtime/issues/4311)

abrown commented on crates/wasmtime/src/runtime/vm/component/libcalls.rs:
```diff
@@ -1267,3 +1267,15 @@ unsafe fn error_context_drop(
+    context: u32,
+) -> Result<u32> {
+    ComponentInstance::from_vmctx(vmctx, |instance| {
+        instance.thread_spawn_indirect(table, element, context)
```
It's unclear to me which parts should live up at this level and which should live in `ComponentInstance::thread_spawn_indirect`. Any suggestions?

alexcrichton replied:
Stylistic an up-to-you,  but generally "simple things" and "ABI things" in this file and meat elsewhere. So e.g. converting `u32` to more typed things here, but leaving the meat of the implementation elsewhere is fine.

alexcrichton commented on crates/cranelift/src/compiler/component.rs:
```diff
@@ -1331,6 +1336,26 @@ impl<'a> TrampolineCompiler<'a> {
+        let element = args[1];
+        let context = args[2];
+
+        // table: u32
+        let table = self
+            .builder
+            .ins()
+            .iconst(ir::types::I32, i64::from(self.offsets.runtime_table(table)));
```
This is a disconnect between runtime and compile-time where here the u32 is the offset in the VMContext but at runtime it's interpreted as `RuntimeTableIndex`

alexcrichton commented on crates/wasmtime/src/runtime/vm/component.rs:
```diff
@@ -745,6 +745,75 @@ impl ComponentInstance {
+        // table.
+        let VMTable { vmctx, from } = self.runtime_table(RuntimeTableIndex::from_u32(table));
+        let element = u64::from(element);
+        let table = unsafe {
+            Instance::from_vmctx(vmctx.as_non_null(), |handle| {
+                let idx = handle.table_index(from.as_non_null().as_ref());
+                handle
+                    .get_defined_table_with_lazy_init(idx, iter::once(element))
+                    .as_ref()
+            })
+        };
```
FWIW technically some and/or most of this is going to go away. Lazy initialization is probably not going to be possible with shared tables.

abrown replied:
Yup. What I'm trying to do is plumb through enough to (a) show how this is currently possible using unshared tables (while still immediately failing) and (b) create a use site here so that future refactoring considers this usage. I was thinking that tests could even check for the initial behavior that _is_ functional today: i.e., we can't run a shared function, but at least we can check that using the wrong index results in a failure.

alexcrichton replied:
I don't think it's possible to hit this at runtime though? All components would fail validation unless they use shared tables, and shared talbes aren't implemented in Wasmtime, so it's not possible to instantiate any component using this intrinsic?

alexcrichton reviewed (commented):
This seems reasonable enough to me with some tweaks, but I do think we'll want a test for this. I don't think there's any viable path to testing the runtime bits at this time so I'd prefer to leave the implementation body of the intrinsic as `todo!()` for now, but testing at least the compile-time/Cranelift parts seems reasonable with a test that just compiles something but does nothing with the result.

abrown committed:
review: do not convert `RuntimeTableIndex` to a context offset too early

abrown committed:
review: plumb through a `TypeFuncIndex` for checking the indirectly-retrieved spawn function

abrown committed:
review: add test

This test cannot pass until more work is done to translate `shared`
types in Wasmtime.
